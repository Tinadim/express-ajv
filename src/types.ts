import { RequestHandler, Request } from 'express';
import Ajv, { Options as AjvOptions } from 'ajv';

export interface RequestSchema {
    type: 'object';
    properties: {
        body?: object;
        params?: object;
        query?: object;
    };
    required?: string[];
    [key: string]: any;
}

export interface ValidatorOptions {
    ajv?: Ajv.Ajv;
    ajvOptions?: AjvOptions;
    formatError?: (error: any) => any;
    getContextParams?: (req: Request) => any;
}

export type SchemaCollection = { [schemaName: string]: RequestSchema };

export interface ValidatorMiddleware {
    ajv: Ajv.Ajv;
    (schemaName: string): RequestHandler;
}
