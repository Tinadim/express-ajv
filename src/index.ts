export {
    SchemaCollection,
    ValidatorMiddleware,
    ValidatorOptions,
} from './types';

export { createValidator } from './create-validator';
