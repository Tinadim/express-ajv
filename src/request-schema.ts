import { RequestSchema } from './types';

function createSchema(querySchema?: object, bodySchema?: object, paramsSchema?: object) {
    const schema: { [key: string]: any } = {
        type: 'object',
        properties: {},
        required: [],
    };

    if (querySchema) {
        schema.required.push('query');
        schema.properties['query'] = querySchema;
    }

    if (bodySchema) {
        schema.required.push('body');
        schema.properties['body'] = bodySchema;
    }

    if (paramsSchema) {
        schema.required.push('params');
        schema.properties['params'] = paramsSchema;
    }

    return schema as RequestSchema;
}

export function createGetRequestSchema(querySchema: object, paramsSchema?: object) {
    return createSchema(querySchema, undefined, paramsSchema);
}

export function createPostRequestSchema(bodySchema: object, paramsSchema?: object) {
    return createSchema(undefined, bodySchema, paramsSchema);
}
