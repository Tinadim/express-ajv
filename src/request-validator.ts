import { forOwn, pick } from 'lodash';
import { RequestHandler, Request, Response, NextFunction } from 'express';
import Ajv, { ValidateFunction } from 'ajv';
import { SchemaCollection, ValidatorOptions } from './types';

export class RequestValidator {
    ajv: Ajv.Ajv;
    private validators: { [schemaName: string]: ValidateFunction } = {};
    private contextExtractor: (req: Request) => any;

    constructor(
        schemas: SchemaCollection,
        options?: ValidatorOptions,
    ) {
        this.ajv = options?.ajv ?? new Ajv(options?.ajvOptions);
        this.contextExtractor = options?.contextExtractor ?? this.defaultValidationContext;
        this.registerSchemas(schemas);
    }

    private registerSchemas(schemaCollection: { [schemaName: string]: object }) {
        forOwn(schemaCollection, (schema: object, schemaName: string) => {
            const validator = this.ajv.compile(schema);
            this.validators[schemaName] = validator;
        });
    }

    private getValidatorForSchema(schemaName: string) {
        if (!schemaName) {
            throw new Error('Please provide the schema name');
        }

        const validator = this.validators[schemaName];
        if (!validator) {
            throw new Error(`Validator for schema with name ${schemaName} not found`);
        }
        return validator;
    }

    private defaultValidationContext(req: Request) {
        return {};
    }

    validateSchema(schemaName: string) {
        const validator = this.getValidatorForSchema(schemaName);

        const middleware: RequestHandler = (req: Request, res: Response, next: NextFunction) => {
            const contextParams = this.contextExtractor(req);
            const context = Object.assign({}, this.ajv, contextParams);
            const dataToValidate = pick(req, ['body', 'params', 'query']);

            return this.validate(dataToValidate, validator, context)
                .then(() => next())
                .catch((error) => {
                    next(error);
                });
        };

        return middleware;
    }

    private validate(data: any, validator: ValidateFunction, context: any) {
        return validator.$async
            ? this.validateAsync(data, validator, context)
            : this.validateSync(data, validator, context);
    }

    private validateSync(data: any, validator: ValidateFunction, context: any) {
        const valid = validator.call(context, data);
        return valid
            ? Promise.resolve()
            : Promise.reject(new Ajv.ValidationError(validator.errors!));
    }

    private validateAsync(data: any, validator: ValidateFunction, context: any) {
        return validator.call(context, data) as Promise<any>;
    }
}
