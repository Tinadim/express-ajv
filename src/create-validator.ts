import { isObject } from 'lodash';
import { SchemaCollection, ValidatorMiddleware, ValidatorOptions } from './types';
import { RequestValidator } from './request-validator';

export function createValidator(schemas: SchemaCollection, options?: ValidatorOptions): ValidatorMiddleware {
    if (!isObject(schemas)) {
        throw new Error('schemas should be an object');
    }

    const validator = new RequestValidator(schemas, options);
    const validatorMiddleware = (schema: string) => validator.validateSchema(schema);
    return Object.assign(validatorMiddleware, { ajv: validator.ajv }) as ValidatorMiddleware;
}
