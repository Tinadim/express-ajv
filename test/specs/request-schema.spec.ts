import { createGetRequestSchema, createPostRequestSchema } from '../../src';

describe('request-schema.spec.ts', () => {
    const baseSchema = {
        type: 'object',
        properties: {
            name: {
                type: 'string',
            },
        },
    };

    const paramsSchema = {
        type: 'object',
        properties: {
            userId: {
                type: 'string',
            },
        },
    };

    it('Should wrap the provided schema with the standard get request schema format without params', () => {
        const result = createGetRequestSchema(baseSchema);
        expect(result).toMatchSnapshot();
    });

    it('Should wrap the provided schema with the standard get request schema format with params', () => {
        const result = createGetRequestSchema(baseSchema, paramsSchema);
        expect(result).toMatchSnapshot();
    });

    it('Should wrap the provided schema with the standard post request schema format without params', () => {
        const result = createPostRequestSchema(baseSchema);
        expect(result).toMatchSnapshot();
    });

    it('Should wrap the provided schema with the standard post request schema format with params', () => {
        const result = createPostRequestSchema(baseSchema, paramsSchema);
        expect(result).toMatchSnapshot();
    });
});
