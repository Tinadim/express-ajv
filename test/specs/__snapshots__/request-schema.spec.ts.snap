// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`request-schema.spec.ts Should wrap the provided schema with the standard get request schema format with params 1`] = `
Object {
  "properties": Object {
    "params": Object {
      "properties": Object {
        "userId": Object {
          "type": "string",
        },
      },
      "type": "object",
    },
    "query": Object {
      "properties": Object {
        "name": Object {
          "type": "string",
        },
      },
      "type": "object",
    },
  },
  "required": Array [
    "query",
    "params",
  ],
  "type": "object",
}
`;

exports[`request-schema.spec.ts Should wrap the provided schema with the standard get request schema format without params 1`] = `
Object {
  "properties": Object {
    "query": Object {
      "properties": Object {
        "name": Object {
          "type": "string",
        },
      },
      "type": "object",
    },
  },
  "required": Array [
    "query",
  ],
  "type": "object",
}
`;

exports[`request-schema.spec.ts Should wrap the provided schema with the standard post request schema format with params 1`] = `
Object {
  "properties": Object {
    "body": Object {
      "properties": Object {
        "name": Object {
          "type": "string",
        },
      },
      "type": "object",
    },
    "params": Object {
      "properties": Object {
        "userId": Object {
          "type": "string",
        },
      },
      "type": "object",
    },
  },
  "required": Array [
    "body",
    "params",
  ],
  "type": "object",
}
`;

exports[`request-schema.spec.ts Should wrap the provided schema with the standard post request schema format without params 1`] = `
Object {
  "properties": Object {
    "body": Object {
      "properties": Object {
        "name": Object {
          "type": "string",
        },
      },
      "type": "object",
    },
  },
  "required": Array [
    "body",
  ],
  "type": "object",
}
`;
